src/main/java/techreborn/blockentity/machine/tier1/ElevatorBlockEntity.java

```diff
@@ -0,0 +1,201 @@
+import reborncore.common.screen.BuiltScreenHandler;
+import reborncore.common.screen.BuiltScreenHandlerProvider;
+import reborncore.common.screen.builder.ScreenHandlerBuilder;
+import techreborn.config.TechRebornConfig;
+import techreborn.init.TRBlockEntities;
+import techreborn.init.TRContent;
+
+import java.util.List;
+import java.util.Optional;
+
+public class ElevatorBlockEntity extends PowerAcceptorBlockEntity implements IToolDrop, BuiltScreenHandlerProvider {
+
+	public static final int MIN_HEIGHT = -64;
+	public static final int MAX_HEIGHT = 319;
+
+	public ElevatorBlockEntity(BlockPos pos, BlockState state) {
+		super(TRBlockEntities.ELEVATOR, pos, state);
+	}
+
+	public static boolean isRunning(final World world, final BlockPos pos) {
+		final BlockEntity entity = world.getBlockEntity(pos);
+		if (!(entity instanceof ElevatorBlockEntity)) {
+			return false;
+		}
+		return ((ElevatorBlockEntity)entity).getStored() > 0;
+	}
+
+	public static boolean isFree(final World world, final BlockPos pos) {
+		return world.getBlockState(pos.up()).isAir() && world.getBlockState(pos.up().up()).isAir();
+	}
+
+	public static boolean isValidTarget(final World world, final BlockPos pos) {
+		return isRunning(world, pos) && isFree(world, pos);
+	}
+
+	public static Optional<BlockPos> nextUpElevator(final World world, final BlockPos pos) {
+		BlockPos upPos = pos.up().up();
+		do {
+			upPos = upPos.up();
+		} while (upPos.getY() <= MAX_HEIGHT && !isValidTarget(world, upPos));
+		if (upPos.getY() < MAX_HEIGHT || isValidTarget(world, upPos)) {
+			return Optional.of(upPos);
+		}
+		return Optional.empty();
+	}
+
+	public static Optional<BlockPos> nextDownElevator(final World world, final BlockPos pos) {
+		BlockPos downPos = pos.down().down();
+		do {
+			downPos = downPos.down();
+		} while (downPos.getY() >= MIN_HEIGHT && !isValidTarget(world, downPos));
+		if (downPos.getY() > MIN_HEIGHT || isValidTarget(world, downPos)) {
+			return Optional.of(downPos);
+		}
+		return Optional.empty();
+	}
+
+	public static int energyCost(BlockPos startPos, BlockPos endPos) {
+		return Math.min(Math.abs(endPos.getY()-startPos.getY())*TechRebornConfig.elevatorEnergyPerBlock,0);
+	}
+
+	protected boolean teleport(World world, BlockPos pos, PlayerEntity player, BlockPos targetPos) {
+		final int energy = energyCost(pos, targetPos);
+		if (getStored() < energy) {
+			return false;
+		}
+		world.playSound(null, pos, SoundEvents.ENTITY_ENDERMAN_TELEPORT, SoundCategory.BLOCKS, 1f, 1f);
+		player.teleport(targetPos.getX(), targetPos.getY(), targetPos.getZ());
+		useEnergy(energy);
+		return true;
+	}
+
+	// PowerAcceptorBlockEntity
+	@Override
+	public void tick(World world, BlockPos pos, BlockState state, MachineBaseBlockEntity blockEntity) {
+		super.tick(world, pos, state, blockEntity);
+		if (world == null || getStored() <= 0 || !isActive(RedstoneConfiguration.POWER_IO)) {
+			return;
+		}
+
+		Optional<BlockPos> upTarget = null;
```

Comment:
I dont think you need both of these, you can just have target. Will let you simplify the code a bit.

Reply:
the idea was when mutliple players teleport at the same time, to only calculate their target once. still holds for sneaking, has been made redundant for jumping